File: part-01/src/test/java/eu/luminis/workshop/smallsteps/UserAppTest.java
package eu.luminis.workshop.smallsteps;

import io.quarkus.test.junit.QuarkusTest;
import org.apache.http.HttpStatus;
import org.junit.jupiter.api.Test;

import static io.restassured.RestAssured.given;
import static org.hamcrest.CoreMatchers.is;

@QuarkusTest
public class UserAppTest {

    @Test
    public void test_missing_content_type() {
        given()
                .when().post("/user")
                .then()
                .statusCode(HttpStatus.SC_UNSUPPORTED_MEDIA_TYPE);
    }

    @Test
    public void test_create_user() {
        given()
                .body("{\"email\": \"harry@example.com\", \"password\": \"mysecret\"}")
                .header("Content-Type", "application/json")
                .when().post("/user")
                .then()
                .statusCode(HttpStatus.SC_CREATED)
                .body(is("User registered correctly"));
    }

    @Test
    public void test_illegal_password() {
        given()
                .body("{\"email\": \"harry@example.com\", \"password\": \"secret\"}")
                .header("Content-Type", "application/json")
                .when().post("/user")
                .then()
                .statusCode(HttpStatus.SC_BAD_REQUEST)
                .body(is("Password must be at least 8 characters long"));

    }
}

File: part-01/src/test/kotlin/eu/luminis/workshop/smallsteps/UserAppTest.kt
package eu.luminis.workshop.smallsteps

import eu.luminis.workshop.smallsteps.api.NewUserRequest
import eu.luminis.workshop.smallsteps.setup.setupUserTestApp
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.testing.*
import kotlin.test.Test
import kotlin.test.assertEquals

class UserAppTest {
    @Test
    fun `should fail when missing contentType`() = testApplication {
        val client = setupUserTestApp()

        client.post("/user").apply {
            assertEquals(HttpStatusCode.UnsupportedMediaType, status)
        }
    }

    @Test
    fun `should create user`() = testApplication {
        val client = setupUserTestApp()

        client.post("/user") {
            contentType(ContentType.Application.Json)
            setBody(
                NewUserRequest(
                    email = "harry@example.com",
                    password = "mysecret"
                )
            )
        }.apply {
            assertEquals(HttpStatusCode.Created, status)
        }
    }

    @Test
    fun `should fail for illegal password`() = testApplication {
        val client = setupUserTestApp()

        client.post("/user") {
            contentType(ContentType.Application.Json)
            setBody(
                NewUserRequest(
                    email = "harry@example.com",
                    password = "secret"
                )
            )
        }.apply {
            assertEquals(HttpStatusCode.BadRequest, status)
            assertEquals("Password must be at least 8 characters long", bodyAsText())
        }
    }
}
